### ISSUE TITLE
Simultaneous calls to add a finalizer and delete can race

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#77988]
Store#Delete begins by looking up the existing object:

https://github.com/kubernetes/kubernetes/blob/8ae998ceb69ae83afe730795aea3bd44913ad868/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L865-L875

If that object has no pending finalizers, and that type does not do graceful deletion, and the incoming deletion options do not specify preconditions (uid or resourceVersion) or GC-related finalizers (like orphan or foreground deletion), then the object is deleted unconditionally:

https://github.com/kubernetes/kubernetes/blob/8ae998ceb69ae83afe730795aea3bd44913ad868/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L929-L932

In between the initial get and the unconditional delete, an update adding finalizers could have been made and succeeded.

/kind bug
/sig api-machinery
/priority important-soon
/cc @caesarxuchao

**Comment by liggitt** [comment]
the checks of the existing object's finalizers should really be done inside a GuaranteedUpdate-like loop, only one that deletes instead of updates

**Comment by ttousai** [comment]
Hello! I'm a bug triage shadow for the 1.16 release cycle and considering this issue is tagged for 1.16, I'd like to check its status. Code freeze is starting on August 29th (about 3 days from now), which means that this PR must merged by then.

Do we still target this issue to be fixed for 1.16?

**Comment by xmudrii** [comment]
@apelisse I'm going to move this issue out of the milestone. If anything changes please let us know.

/milestone clear

**Comment by markyjackson-taulia** [comment]
@apelisse Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

**Comment by tedyu** [comment]
In Store#updateForGracefulDeletionAndFinalizers (called by Store#Delete), I see e.Storage.GuaranteedUpdate call wrapping the SimpleUpdate func.

**Comment by apelisse** [comment]
I've had 0 time to look at this, if you want to take a shot at it @tedyu, let me know if I can help?

**Comment by markyjackson-taulia** [comment]
A kind reminder that code freeze for this release is on November 14. Is this issue still intended for 1.17? If by weeks end we have no traction I will have to move to 1.18.

**Comment by apelisse** [comment]
We're not looking at this, you can move it to 1.18

**Comment by annajung** [comment]
Hello @apelisse  👋 I'm one of the Bug Triage shadows for the 1.18 release. Since this issue is tagged for 1.18, I'd like to check its status. Is this issue intended for milestone 1.18? 

Friendly reminder: Code freeze is scheduled for March 5th.

**Comment by apelisse** [comment]
No :-/

**Comment by annajung** [comment]
@apelisse Thanks for getting back to me! Since this issue is not intended for 1.18 release, would you be okay if I defer the milestone to 1.19?

**Comment by apelisse** [comment]
Of course

**Comment by bai** [comment]
👋 Hello from Bug Triage team! Wanted to follow up on this issue with a friendly reminder that the code freeze for 1.19 is starting June 25th (about 4 weeks from now).

As this issue is tagged for 1.19, is it still planned for this release?

**Comment by sschmitt** [comment]
Is there a recommended procedure to implement on the controller side to avoid this race when adding a finalizer?
Would something like this avoid the race?
1. Add finalizer patch/update
2. Get object from API server
3. If get succeeds, continue (it's now safe to create external artifacts?)

**Comment by liggitt** [comment]
Only slightly related, but I wanted to capture other bits about the interaction of the delete flow, admission, and  around preconditions and admission from a [slack conversation](https://kubernetes.slack.com/archives/C0EG7JC6T/p1711478299810659):


tl;dr:
1. delete admission is accurately called with the correct old object before the calling API server modifies/deletes the object (either to set deletionTimestamp for graceful deletion, or to set a finalizer for GC deletion, or to actually remove the object from etcd). There's an edge where the API server reports success for a Delete call because the underlying object is already gone where delete admission won't be called with a nil oldObject, but... it wasn't actually guarding an action taken by the API server anyway, so /shrug
2. The API server tries to optimize away get calls to etcd by using a cached copy of the object as the "old object". If the cached copy is stale, that becomes evident in the conditionalDelete and the latest object is fetched from etcd, and admission is re-called.
3. All the modifications / deletions the API server does are revision-qualified (even if the triggering API request wasn't) and will retry if there's a race with a concurrent update, which will call admission again.
4. There's a long-standing bug (https://github.com/kubernetes/kubernetes/issues/77988) where existence of finalizers is not checked inside the conditional retry loop in the API server, so it will actually allow deletion of an object that concurrently had a finalizer added to it, but admission will actually get called and will see that object.

---

Long version:

https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go#L129 is where admission gets wired into delete request handling

AdmissionToValidateObjectDeleteFunc returns a closure that gets invoked by the apiserver storage Delete handler with current `old` object as seen by the apiserver - https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go#L172-L179

https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1096 is the Delete handler. It is ... twisty ...

There's two ways delete admission can get called from inside that method:
1. https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1140C65-L1140C103 → https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1013-L1020

`updateForGracefulDeletionAndFinalizers` is called in cases where we are admitting a delete request but expect to set deletionTimestamp and maybe not actually remove the object from etcd in this handler yet (for one of these reasons: it has finalizers present, or is a "graceful deletion" object and the delete is not a force-delete, or the delete request indicated it wanted async deletion behavior where child objects are deleted first ... essentially an "add finalizer, delete children, come back and remove finalizer later" opt-in). Here, delete admission gets called inside a GuaranteedUpdate loop that ensures oldObject is the latest content from etcd. If the GuaranteedUpdate races with an update from another API server, it gets called again with the updated old object (admission can be called multiple times for a single API request for this reason).

2. https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1172 → https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L257 → https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L289-L350

This path is called when we are actually deleting the object from etcd. Here, delete admission gets called inside a for loop that gets current state from etcd, checks preconditions (if any), calls delete admission, then does a conditional delete from etcd. If the conditional delete races with an *update* from another API server, it gets called again with the updated old object (admission can be called multiple times for a single API request for this reason).

If the conditional delete races with a concurrent *delete*, it is possible the conditional delete fails with an error and returns, and (depending on the resource type), the overall delete API request reports success (https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1175-L1179)

**Comment by liggitt** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/129768 as a proof-of-concept of a solution to this if someone wants to pick that up and add tests, etc

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/129768 is ready for review

**Comment by liggitt** [comment]
Fix was reverted in https://github.com/kubernetes/kubernetes/pull/133979, reopening
